### ISSUE TITLE
Flickering issues with function drawing examples

### ISSUE THREAD

**Issue by einarf** [pythonarcade/arcade#858]
The following examples flicker (at least on windows for me):

```
arcade/examples/nested_loops_top_right_triangle.py
arcade/examples/nested_loops_top_left_triangle.py
arcade/examples/nested_loops_box.py
arcade/examples/nested_loops_bottom_right_triangle.py
arcade/examples/nested_loops_bottom_left_triangle.py
arcade/examples/happy_face.py
arcade/examples/drawing_with_loops.py
arcade/examples/drawing_with_functions.py
arcade/examples/drawing_primitives.py
```

By flickering I mean that the window swaps between a black screen and the rendered result. This happens based on your your OS or enviroment. It all relies on how the framebuffer is updated internally.

A possible solution solution is to make `arcade.open_window` create a single buffered window or we can add a parameter for doube or single buffering.

```py
arcade.open_window(600, 600, "Drawing Primitives Example", double_buffer=False)
```

I'm pretty sure users might also experince this going through the arcade book? (Depending on their enviroment) @pvcraven

**Comment by einarf** [comment]
This is actually a pyglet 2 think at least on windows. False alarm for 2.5.x. It's probably related to how the event loop changed.

**Comment by einarf** [comment]
How did this turn out? I have an alternative solution if the current one is not reliable enough.

**Comment by einarf** [comment]
Another problem is: If the minimize the window and show it again the window is not blank:

I have a possible solutiuon for this:

* Check if the event dispatcher don't have events in `window._event_stack`
* Check if `window.on_draw` in the the default `on_draw` in `arcade.Window`

If these two contions are true when `start_render` is called we know that the user is doing simple module level rendering with `start/finish_render`. We can use this to capture what is rendered between start/finish_render into a `Framebuffer` and keep displaying the contents of this frambuffer every frame.

We can monkeypatch `start_render` and `window.on_draw` for this.

**Comment by pvcraven** [comment]
This seemed fixed with the prior code changes.

**Comment by J3ronimo** [comment]
I see the same effect (arcade 2.6.17, Python 3.9, Win10, RTX2070, 144Hz monitor).

Thought it was my code at first, but then saw the same in the examples, like this one
> arcade\examples\gui_scrollable_text.py

It's very slight, but the sceen seems to flicker into black for a frame or so in between, a couple of times per second.
Although in my case, `nested_loops_bottom_left_triangle.py` does not flicker, which renders only once. So I suppose it must have to do with `arcade.start_render()` (or `self.clear()`) and the flipping of buffers.

**Comment by einarf** [comment]
> It's very slight, but the sceen seems to flicker into black for a frame or so in between, a couple of times per second.

Yup. It's because we're trying to trick the double buffering in the window skipping every other frame and displaying only the one we rendred into. This worked great in pyglet 1.5, but in pyglet 2 the event system changed (for the better in most other cases).

I think the only way to fix this is making a separate system for doing this kind of "one frame" rendering. It would involve a framebuffer object we draw to and we just keep presenting this framebuffere every frame. However.. it will probably break all existing examples using this method.

Framebuffers can be single, double and tripple buffered so the current way doesn't work properly in like 25% of instances. If you minimize or resize the window the content is also removed.

**Comment by einarf** [comment]
I'm seeing this issue as well. I think we might need to re-think how to solve this fully. The changes @pvcraven made probably made things 80% better, but I think as people get newer hardware we'll keep getting more and more people with problems.

I could try to mock up something using Framebuffer object. Maybe we can also get rid of `start/finish_render`. It must be something extremely simple or at least not anything more complicated than the current setup.

We could either remove or deprecate the old solution depending on how feasable it is to keep it in.

**Comment by einarf** [comment]
I think that's happening is:
* Some systems are operating with tripple buffering. The 3rd buffer is not always used, but it can cycle in depeding in internal conditions. It's nothing we can control
* Some systems have double buffering, but the "flip" is blitting the back buffer to the front buffer.
* The `on_draw`/`on_update` intervals are not 100% reliable. You can have two updates before a draw or two draw before an update in some cases.
* On most systems a window resize or show/hide will always destroy the window contents completely. There is no way for us currently to recover from this since we render the contents once.

Rendering to a Framebuffer and simply dispaying rendering this content every frame should solve all these issues.

**Comment by gran4** [comment]
> I'm seeing this issue as well. I think we might need to re-think how to solve this fully. The changes @pvcraven made probably made things 80% better, but I think as people get newer hardware we'll keep getting more and more people with problems.

I saw it as well in my game when I resized the window.(arcade 2.6.17, Mac 2017, Monterey)

> I think as people get newer hardware we'll keep getting more and more people with problems.

My hardware isn't that new. It's 6 years old. Also, it is a Mac not a windows device. Is this a separate problem?

**Comment by einarf** [comment]
> My hardware isn't that new. It's 6 years old. Also, it is a Mac not a windows device. Is this a separate problem?

It's a separate problem. Your issue might be related to vsync. It tends to reset after resize.

**Comment by DragonMoffon** [comment]
I don't know if arcade.finish_render() still exists, but we could make it that this method saves the current screen and redraws it until start_render is called again.

Or maybe a new method save_render()? capture_render()?

**Comment by einarf** [comment]
The way to completely solve this is making the user somehow render to a framebuffer object. We simply just render this framebuffer to the screen each frame. However, this is not an easy task since `start_render()` is both used window classes and in simple module based snippets.

**Comment by einarf** [comment]
This is finally fixed in 3.0
